pkg/wormhole/client/health.go

```diff
@@ -0,0 +1,166 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"sync/atomic"
+	"time"
+
+	kitlog "github.com/go-kit/kit/log"
+	"golang.org/x/net/ipv4"
+	"k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kubernetes_clientset "k8s.io/client-go/kubernetes"
+
+	"github.com/sapcc/kubernikus/pkg/client/kubernetes"
+	"github.com/sapcc/kubernikus/pkg/util/icmp"
+	"github.com/sapcc/kubernikus/pkg/util/netutil"
+)
+
+var sequenceNumber uint32 = 0
+
+const checkInterval = 60 * time.Second
+const intervalJitter = 0.5
+
+const NodeRouteBroken v1.NodeConditionType = "RouteBroken"
+
+type healthChecker struct {
+	nodeName string
+	localIP  net.IP
+	listener *icmp.Listener
+	logger   kitlog.Logger
+	client   kubernetes_clientset.Interface
+}
+
+func NewHealthChecker(kubeconfig, context, nodeNameOverride string, logger kitlog.Logger) (*healthChecker, error) {
+	var hc = healthChecker{logger: logger, nodeName: nodeNameOverride}
+	var err error
+
+	if hc.nodeName == "" {
+		if hc.nodeName, err = os.Hostname(); err != nil {
+			return nil, fmt.Errorf("node name not given and hostname unvailable")
+		}
+	}
+
+	if hc.client, err = kubernetes.NewClient(kubeconfig, context, logger); err != nil {
+		return nil, fmt.Errorf("Failed to create kubernetes client: %s", err)
+	}
+
+	interfaceName, err := netutil.DefaultInterfaceName()
+	if err != nil {
+		return nil, err
+	}
+
+	if hc.localIP, err = netutil.InterfaceAddress(interfaceName); err != nil {
+		return nil, err
+	}
+
+	return &hc, nil
+}
+
+func (hc *healthChecker) Start(stopCh <-chan struct{}) error {
+	hc.logger.Log("msg", "Starting healthecker", "interval", checkInterval, "jitter", intervalJitter)
+
+	var err error
+	if hc.listener, err = icmp.NewListener(hc.localIP.String()); err != nil {
+		return err
+	}
+
+	wait.JitterUntil(hc.Reconcile, checkInterval, intervalJitter, true, stopCh)
+
+	return nil
+}
+
+func (hc *healthChecker) Reconcile() {
+	if err := hc.reconcile(); err != nil {
+		hc.logger.Log("msg", "healthcheck", "err", err)
+	}
+}
+
+func (hc *healthChecker) reconcile() error {
+
+	destinationIP, err := netutil.InterfaceAddress("cbr0")
```

Comment:
Yes. Instead of doing this we could also use the first ip of the pod cidr range. I could look into that. There is one caveat. If no pod has been scheduled on a node yet, the interface is not created yet. Currently this leads to an early exit of the healthcheck without setting the condition. If we switch to using the the first`podCIDR` IP we would starting pining a non-existing ip. I'm not sure if the router continues to send redirects in that case.

Reply:
Leave it until we need to make it more complicated